src/packages/contrib/contrib.cc

```diff
@@ -1610,116 +1610,131 @@ void f_replaceable(void) {
-        ignore[i + 2] = 0;
-      }
-    }
-    numignore += 2;
     obj = (sp - 1)->u.ob;
-  } else {
-    numignore = 2;
-    ignore = reinterpret_cast<char **>(DCALLOC(2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
-    ignore[0] = findstring(APPLY_CREATE);
-    ignore[1] = findstring(APPLY___INIT);
+  }
+  else {
     obj = sp->u.ob;
   }
 
   prog = obj->prog;
-  num = prog->num_functions_defined + prog->last_inherited;
-
-  for (i = 0; i < num; i++) {
-    if (prog->function_flags[i] & (FUNC_INHERITED | FUNC_NO_CODE)) {
-      continue;
+  if (obj == simul_efun_ob || prog->func_ref) {
+    replaceable = 0;
+  }
+  else {
+    if (st_num_arg == 2) {
+      numignore = sp->u.arr->size;
+      if (numignore) {
+        ignore = reinterpret_cast<char **>(
+            DCALLOC(numignore + 2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
+      } else {
+        ignore = 0;
+      }
+      ignore[0] = findstring(APPLY_CREATE);
+      ignore[1] = findstring(APPLY___INIT);
+      for (i = 0; i < numignore; i++) {
+        if (sp->u.arr->item[i].type == T_STRING) {
+          ignore[i + 2] = findstring(sp->u.arr->item[i].u.string);
+        } else {
+          ignore[i + 2] = 0;
+        }
+      }
+      numignore += 2;
+    } else {
+      numignore = 2;
+      ignore = reinterpret_cast<char **>(DCALLOC(2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
+      ignore[0] = findstring(APPLY_CREATE);
+      ignore[1] = findstring(APPLY___INIT);
     }
-    for (j = 0; j < numignore; j++) {
-      if (ignore[j] == find_func_entry(prog, i)->funcname) {
+
+    num = prog->num_functions_defined + prog->last_inherited;
+
+    for (i = 0; i < num; i++) {
+      if (prog->function_flags[i] & (FUNC_INHERITED | FUNC_NO_CODE)) {
+        continue;
+      }
+      for (j = 0; j < numignore; j++) {
+        if (ignore[j] == find_func_entry(prog, i)->funcname) {
+          break;
+        }
+      }
+      if (j == numignore) {
         break;
       }
     }
-    if (j == numignore) {
-      break;
-    }
-  }
 
-  replaceable = (i == num);
-  if (obj == simul_efun_ob || prog->func_ref) {
-    replaceable = 0;
+    replaceable = (i == num);
+    FREE(ignore);
   }
 
   if (st_num_arg == 2) {
     free_array((sp--)->u.arr);
   }
-  FREE(ignore);
   free_svalue(sp, "f_replaceable");
   put_number(replaceable);
 }
 #endif
 
 #ifdef F_PROGRAM_INFO
+typedef struct {
+    int h_c,    // how many headers?
+        p_s,    // how big are the compiled programs?
+        ff_c,   // how many functions flags?
+        fd_c,   // how many function definitions?
+        s_c,    // how many strings?
+        v_c,    // how many variables?
+        i_c,    // how many inherits?
+        nc_c,   // how many classes (structs)?
+        mc_c,   // how many class members?
+        t_c,    // how mamy saved types (#pragma save_types)?
+        total;  // how big all together?
+} program_info_t;
+
+inline void sum_program_info(const program_t * const p, program_info_t &info) {
```

Comment:
_t has been used, by C programmers, to indicate typedef... at least as far
back as when I was in school, in the late 1980's.  You certainly could
invent your own thing for it, but nobody but YOU will not assume it means
typedef. :)

It was also not uncommon for people to use _s to indicate variables that
were a structure.  The use of both standards made it clear what code was
doing without needing to go look at the header files to see.

For example:

struct foo_s { ... };
struct foo_s foo;
...
foo.blah = 1;

typedef struct { ... } foo_t;
foo_t stuff;

stuff.blah = 1;

If the struct or typedefs are in header files away from the declarations of
the variables that use them, it makes it easier to FIND said declarations.
Not as big a deal if you're using a smart IDE (visual studio), but helpful
for folks using vim and grep.


On Thu, Jan 7, 2016 at 1:55 PM, Shea690901 <notifications@github.com> wrote:

> In src/packages/contrib/contrib.cc
> <https://github.com/fluffos/fluffos/pull/297#discussion_r49110355>:
>
> >    free_svalue(sp, "f_replaceable");
> >    put_number(replaceable);
> >  }
> >  #endif
> >
> >  #ifdef F_PROGRAM_INFO
> > +typedef struct {
> > +    int h_c,    // how many headers?
> > +        p_s,    // how big are the compiled programs?
> > +        ff_c,   // how many functions flags?
> > +        fd_c,   // how many function definitions?
> > +        s_c,    // how many strings?
> > +        v_c,    // how many variables?
> > +        i_c,    // how many inherits?
> > +        nc_c,   // how many classes (structs)?
> > +        mc_c,   // how many class members?
> > +        t_c,    // how mamy saved types (#pragma save_types)?
> > +        total;  // how big all together?
> > +} program_info_t;
> > +
> > +inline void sum_program_info(const program_t * const p, program_info_t &info) {
>
> You mean one can use struct ident {}; and then later ident variable; in
> C++????
> At least as long as ident is uniqe…
> *That* I've to test…
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fluffos/fluffos/pull/297/files#r49110355>.
>